specs/core/1_new_shards.md

```diff
@@ -0,0 +1,440 @@
+        for i, v in zip(committee, validators)
+    ]
+    pubkeys = [v.pubkey for v in validators]
+    return CompactCommittee(pubkeys=pubkeys, compact_validators=compact_validators)
+```
+
+### `get_light_client_committee`
+
+```python
+def get_light_client_committee(beacon_state: BeaconState, epoch: Epoch) -> Sequence[ValidatorIndex]:
+    assert epoch % LIGHT_CLIENT_COMMITTEE_PERIOD == 0
+    active_validator_indices = get_active_validator_indices(beacon_state, epoch)
+    seed = get_seed(beacon_state, epoch, DOMAIN_SHARD_LIGHT_CLIENT)
+    return compute_committee(active_validator_indices, seed, 0, ACTIVE_SHARDS)[:TARGET_COMMITTEE_SIZE]
+```
+
+### `get_indexed_attestation`
+
+```python
+def get_indexed_attestation(beacon_state: BeaconState, attestation: Attestation) -> IndexedAttestation:
+    attesting_indices = get_attesting_indices(state, attestation.data, attestation.aggregation_bits)
+    return IndexedAttestation(attesting_indices, data, custody_bits, signature)
+```
+
+### `update_gasprice`
+
+```python
+def update_gasprice(prev_gasprice: Gwei, length: uint8) -> Gwei:
+    if length > BLOCK_SIZE_TARGET:
+        delta = prev_gasprice * (length - BLOCK_SIZE_TARGET) // BLOCK_SIZE_TARGET // GASPRICE_ADJUSTMENT_COEFFICIENT
+        return min(prev_gasprice + delta, MAX_GASPRICE)
+    else:
+        delta = prev_gasprice * (BLOCK_SIZE_TARGET - length) // BLOCK_SIZE_TARGET // GASPRICE_ADJUSTMENT_COEFFICIENT
+        if delta > prev_gasprice - GASPRICE_ADJUSTMENT_COEFFICIENT:
+            return GASPRICE_ADJUSTMENT_COEFFICIENT
+        else:
+            return prev_gasprice - delta
+```
+
+### `is_valid_indexed_attestation`
+
+```python
+def is_valid_indexed_attestation(state: BeaconState, indexed_attestation: IndexedAttestation) -> bool:
+    """
+    Check if ``indexed_attestation`` has valid indices and signature.
+    """
+
+    # Verify indices are sorted
+    if indexed_attestation.participants != sorted(indexed_attestation.participants):
+        return False
+    
+    # Verify aggregate signature
+    all_pubkeys = []
+    all_message_hashes = []
+    for participant, custody_bits in zip(participants, indexed_attestation.custody_bits):
+        for i, bit in enumerate(custody_bits):
+            all_pubkeys.append(state.validators[participant].pubkey)
+            # Note: only 2N distinct message hashes
+            all_message_hashes.append(AttestationCustodyBitWrapper(hash_tree_root(indexed_attestation.data), i, bit))
+        
+    return bls_verify_multiple(
+        pubkeys=all_pubkeys,
+        message_hashes=all_message_hashes,
+        signature=indexed_attestation.signature,
+        domain=get_domain(state, DOMAIN_BEACON_ATTESTER, indexed_attestation.data.target.epoch),
+    )
+```
+
+## Beacon Chain Changes
+
+### New state variables
+
+```python
+    shard_states: Vector[ShardState, MAX_SHARDS]
+    online_countdown: Bytes[VALIDATOR_REGISTRY_LIMIT]
```

Comment:
I was keeping it as a list of bytes because that way the hashing is lower. Hashing over a uint64 for every validator is currently only done in the validator balances list and is a primary component of epoch transition overhead; adding yet more hashing of this type would significantly add overhead.

If we think using bytes/uint8 is too impure, I have another suggestion: have a bitlist "online during previous 4-epoch period" and "online during current 4-epoch period". Validators would be considered online if their bit if flipped on for either one. This would have even less hashing overhead and would use existing types.

Reply:
I think List[uint8, ...] is clearer for this purpose than Bytes[...]